karapace/backup/api.py

```diff
@@ -386,17 +388,28 @@ def _handle_producer_send(
         instruction.key,
         instruction.value,
     )
-    try:
-        producer.send(
-            instruction.topic_name,
-            key=instruction.key,
-            value=instruction.value,
-            partition=instruction.partition_index,
-            headers=[(key.decode() if key is not None else None, value) for key, value in instruction.headers],
-            timestamp=instruction.timestamp,
-        ).add_done_callback(producer_callback)
-    except (KafkaError, AssertionError) as exc:
-        raise BackupDataRestorationError("Error while calling send on restoring messages") from exc
+    try_sending = True
+    tries = 0
+    while try_sending:
+        try:
+            tries += 1
+            producer.send(
+                instruction.topic_name,
+                key=instruction.key,
+                value=instruction.value,
+                partition=instruction.partition_index,
+                headers=[(key.decode() if key is not None else None, value) for key, value in instruction.headers],
+                timestamp=instruction.timestamp,
+            ).add_done_callback(producer_callback)
```

Comment:
issue: Can there be odd side-effects of registering the same callback for multiple operations? It's not clear to me how this will behave.

Reply:
I moved the callback adding to the `else`.